AI #2373: Add `Data Generator Name` to `Cost and Usage` dataset (#2376)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/datageneratorname.md

# Data Generator Name

> Data Generator Name is the name of the entity, platform, or system responsible for constructing and delivering a [*dataset artifact*](#glossary:dataset-artifact). It is commonly used to establish data provenance and traceability for multi-provider analysis, explicitly separating the entity providing services from the entity providing data.

Reviewer: The existing [Data Generator](#metadata.datagenerator) metadata property describes the same concept at the file level ("Human-readable name of the entity that generated the dataset instance"). Promoting it to a per-row dimension implies a generator-side single-value constraint (column = `metadata.DataGenerator` for every row at generation), with the column's value-add being preservation of per-row provenance through downstream practitioner unions of multiple dataset instances. Worth adding explicit language in the column file to that effect, so generators don't read "Mandatory + no nulls" as permission to vary the value within a file at generation.

Separately, the sibling Name columns (`ServiceProviderName`, `HostProviderName`, `InvoiceIssuerName`) all use "entity" while this column expands to "entity, platform, or system". Worth a quick check on whether that's intentional differentiation or drift.

Author: How about:

```suggestion
Data Generator Name is the name of the entity responsible for constructing and delivering a [*dataset artifact*](#glossary:dataset-artifact). It preserves per-row data provenance and traceability when practitioners union multiple *dataset artifacts* for multi-provider analysis, explicitly separating the entity providing services from the entity providing data.
```

## Requirements

DataGeneratorName MUST adhere to the following requirements:

* DataGeneratorName MUST be of type String.
* DataGeneratorName MUST conform to [StringHandling](#attributes.stringhandling) requirements.
* DataGeneratorName MUST NOT be null.

> * DataGeneratorName MUST represent the specific system, tool, or entity that generated the *dataset artifact*.

Reviewer: ```suggestion
* DataGeneratorName MUST represent the entity, platform, or system that generated the *dataset artifact*.
```

The opening paragraph and Description describe the actors as "entity, platform, or system." Aligning the Requirements bullet to match makes the column file's three sections consistent. No normative content change.

Author: I think that's good.  Further, we need to double-down on the fact that this is a single entity across the entire delivery, per your feedback above.  How about:

```suggestion
* DataGeneratorName MUST represent the entity that generated the *dataset artifact*.
* DataGeneratorName MUST contain the identical value for every row within the *dataset artifact*.
```

## Column ID

DataGeneratorName

## Display Name

Data Generator Name

## Description

> The name of the entity, platform, or system responsible for generating a *dataset artifact*.

Author: @Matt-Cowsert
Per the above, how about:

```suggestion
The name of the entity responsible for constructing and delivering a *dataset artifact*.
```
